streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -342,7 +342,19 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
         maybeThrowTaskExceptions(taskCloseExceptions);
 
-        createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        rebalanceInProgress = true;
+        final Collection<Task> newActiveTasks = createNewTasks(activeTasksToCreate, standbyTasksToCreate);
+        // If there are any transactions in flight and there are newly created active tasks, commit the tasks
+        // to avoid potential long restoration times.
+        if (processingMode == EXACTLY_ONCE_V2 && threadProducer().transactionInFlight() && !newActiveTasks.isEmpty()) {
+            log.info("New active tasks were added and there is an inflight transaction. Attempting to commit tasks.");
+            final int numCommitted = commitTasksAndMaybeUpdateCommittableOffsets(tasks.allTasks(), new HashMap<>());
```

Comment:
We pass in the second parameter to be able to handle timeout exceptions correctly (cf. other places where we call `commitTasksAndMaybeUpdateCommittableOffsets`) -- here, we don't catch `TimeoutException`. What is the impact? Seems it bubble into the consumer and crashes us?

Given that we kinda need to commit, it seems there are two things we could do: either add a loop here, and retry the commit until we exceed `task.timeout.config` (not my personally preferred solution), or actually move the whole commit logic into the restore part (not sure if this is easily possible) -- ie, each time we enter the restore code, we check if there is an open TX and commit. Not sure how this would align this the new state-updated core though.

\cc @cadonna @lucasbru -- can you comment?

Reply:
In the state updater code path, we would immediately go back to processing / committing the active (previously existing) tasks, so I am not sure the problem being solved here may not be relevant in the state updater code. But I may be missing something. 

In that case, we should probably only execute the commit if the state updater code path is not enabled.

Reply:
Thanks @lucasbru. Are you suggesting that with new state-updated core, we won't get the problems which this PR is trying to solve? I am not aware of the the new state updater code path TBH.

Reply:
In the state updater code path, the restoration for the new task will be done by a separate thread, and the main thread will immediately go back to processing. There is an internal configuration (StreamConfig.InternalConfig) to enable the state updater, if you have an integration test for this bug you could try it with the flag enabled.

Reply:
@lucasbru thanks for the input. I don't have an IT for this because it was testable via unit tests. TBH, this PR is pretty old and lot of things have changed in the streams runtime since then and I am not totally upto date with those. Will need to revisit.